### DISCUSSION TITLE
Per-file variables via frontmatter key `variables`

### DISCUSSION THREAD

**Discussion by rben01** [quarto-dev/quarto-cli#8947] [Feature Requests]
### Description

Currently, Quarto files can include [variables](https://quarto.org/docs/authoring/variables.html#meta) in one of two ways: `{{< var foo >}}` with a project-wide `_variables.yml`, or `{{< meta bar >}}` from the file’s metadata. It would be great to have file-local variables with a `variables` key in the frontmatter that could be used like `{{< meta variables.baz >}}` (or perhaps a shorthand like `{{< filevar baz >}}`).

**Comment by mcanouil** [comment]
If you set the variables in the frontmatter, it means those are metadata so you can use the metadata shortcode to retrieve them.

If you want to have a "variables" key, then I suggest to write a Lua filter/shortcode for this purpose.

I don't think there is something to be done in Quarto here, but I might have missed the purpose of your request and in which way Quarto does not achieve that currently.

**Reply by rben01** [reply]
I guess it would just be for organization and documentation — it would demonstrate that the variable isn't really document metadata (like title and authors and the like), but only exists to be substituted into the document body.

**Reply by mcanouil** [reply]
Anything but "params" (which comes from `knitr` not Quarto) in the frontmatter is metadata. That's how it is and it's very unlikely going to change.
As I said, now if you want to add some specific definition such as "variables" in addition to be a metadata (you can't change that) is a "variable" you'll have to make that yourself.
Making something like that inside Quarto will only create confusion as what are variables and what are metadata.
Feel free to create your own metadata organisational convention such as:

```md
---
my-variables:
  my-first-variable: whatever
---

{{< meta my-variables.my-first-variable >}}
```

Or create a Lua filter/shortcode named "filevar" which will basically read the metadata and read only "variables".

```md
---
my-variables:
  my-first-variable: whatever
---

{{< filevar my-first-variable >}}

or

{{< meta my-variables.my-first-variable >}}
```

See <https://github.com/mcanouil/quarto-lua-env> which adds metadata from the Lua environment which can then be accessed using the following two approaches:

```md
{{< lua-env quarto.doc.input_file >}}

{{< meta lua-env.quarto.doc.input_file >}}
```

**Reply by rben01** [reply]
I tried your first option, but it didn't work, which is why I opened this issue.

> WARNING (/Applications/quarto/share/filters/main.lua:15686) Unknown meta key variables.my-first-variable specified in a metadata Shortcode.

The result in the output is `?meta:variables.my-first-variable`.

**Reply by mcanouil** [reply]
It appears `variables` is used internally to provide "Pandoc metadata variables" so you cannot use it.

**Reply by rben01** [reply]
Oh, lol. I opened this discussion precisely because I tried to use `variables` as the grouping key and it didn't work. Switching to `vars` does indeed work. Thanks!

Should this behavior be documented somewhere?

**Reply by mcanouil** [reply]
Rather than documentation, I believe the best option is YAML validation to warn users.
